build-tools/xaprepare/xaprepare/Steps/Step_GenerateCGManifest.cs

```diff
@@ -0,0 +1,302 @@
+			"mono/debugger-libs",
+			"Newtonsoft.Json",
+			"System.CommandLine",
+			"Xamarin.Forms",
+			"xunit",
+			"xunit.abstractions",
+			"xunit.analyzers",
+			"xunit.assert",
+			"xunit.core",
+			"xunit.extensibility.core",
+			"xunit.extensibility.execution",
+			"xunit.runner.utility",
+
+			// ???
+			"Microsoft.Build",
+			"Microsoft.Build.Framework",
+			"Microsoft.Build.Tasks.Core",
+			"Microsoft.Build.Utilities.Core",
+			"Microsoft.VisualStudio.CoreUtility",
+			"Microsoft.VisualStudio.Imaging",
+			"Microsoft.VisualStudio.OLE.Interop",
+			"Microsoft.VisualStudio.Shell.15.0",
+			"Microsoft.VisualStudio.Shell.Framework",
+			"Microsoft.VisualStudio.Shell.Interop",
+			"Microsoft.VisualStudio.Shell.Interop.8.0",
+			"Microsoft.VisualStudio.Shell.Interop.9.0",
+			"Microsoft.VisualStudio.TextManager.Interop",
+			"Microsoft.VisualStudio.TextManager.Interop.8.0",
+			"Microsoft.VisualStudio.Threading",
+			"Microsoft.VisualStudio.Utilities",
+			"Microsoft.VisualStudio.Validation",
+			"Microsoft.VSSDK.BuildTools",
+		};
+
+		public Step_GenerateCGManifest ()
+			: base ("Generate CGManifest.json")
+		{}
+
+		protected override async Task<bool> Execute (Context context)
+		{
+			var nugets              = MSBuildPackageReferenceInfo.GetPackageReferences ();
+			var git                 = new GitRunner (context);
+			var gitSubmoduleInfo    = await git.ConfigList (new[]{"--blob", "HEAD:.gitmodules"});
+			var gitSubmoduleStatus  = await git.SubmoduleStatus ();
+			var gitSubmodules       = GitSubmoduleInfo.GetGitSubmodules (gitSubmoduleInfo, gitSubmoduleStatus);
+
+			var cgManifestEntries   = ((IEnumerable<CGManifestEntry>) nugets).Concat (gitSubmodules)
+				.OrderBy (e => e.Name);
+
+			var jsonPath    = Path.Combine (Configurables.Paths.BuildBinDir, "CGManifest.json");
+			using var json  = File.CreateText (jsonPath);
+
+			json.WriteLine ("{");
+			json.WriteLine ("    \"Registrations\": [");
+
+
+			bool first = true;
+
+			foreach (var entry in cgManifestEntries) {
+				if (!first) {
+					json.WriteLine (",");
+				}
+				first = false;
+				entry.WriteJsonTo (json, DevelopmentDependencies);
+			}
+
+			json.WriteLine ();
+			json.WriteLine ("    ]");
+			json.WriteLine ("}");
+
+			return true;
+		}
+	}
+
+	abstract class CGManifestEntry {
+
+		public  abstract    string  Name {get;}
+		public  abstract    void    WriteJsonTo (StreamWriter writer, HashSet<string> developmentDependencies);
+
+		protected CGManifestEntry ()
+		{
+		}
+	}
+
+	sealed class GitSubmoduleInfo : CGManifestEntry {
+
+		public  override    string      Name {
+			get {
+				const string github = "github.com/";
+				int i = RepositoryUrl.IndexOf (github, StringComparison.OrdinalIgnoreCase);
+				if (i >= 0)
+					return RepositoryUrl.Substring (i + github.Length);
+				return RepositoryUrl;
+			}
+		}
+
+		public  string      RepositoryUrl   {get; private set;}
+		public  string      CommitHash      {get; private set;}
+
+		private GitSubmoduleInfo ()
```

Comment:
`private` is the default visibility for all class members, no need to us the keyword